File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.cpp

class compare
{
public:
    bool operator()(const pair<int, int> &i, const pair<int, int> &r)
    {

        return i.second < r.second;
    }
};

class Solution
{
public:
    vector<int> topKFrequent(vector<int> &nums, int k)
    { // number --> freq
        unordered_map<int, int> mp;
        priority_queue<pair<int, int>, vector<pair<int, int>>, compare> q;
        vector<int> res;
        for (auto i : nums)
        {
            mp[i]++;
        }

        for (auto i : mp)
        {
            q.push(i);
        }

        for (int i = 0; i < k; i++)
        {
            res.push_back(q.top().first);
            q.pop();
        }

        return res;
    }
};

File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.java

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;

class Solution {
    public int[] topKFrequent(int[] nums, int k) {
        Map<Integer, Integer> countMap = new HashMap<>();
        int [] res = new int [k];
        Arrays.stream(nums).forEach(num -> countMap.put(num, countMap.getOrDefault(num, 0) + 1));

        ArrayList<Integer>[] buckets = new ArrayList[nums.length + 1]; // array of arrayList ---> key is frequency and value is array list  of repeated objects in this length
        for (int key : countMap.keySet()) {
            int freq = countMap.get(key);
            if (buckets[freq] == null)
                buckets[freq] = new ArrayList<>();
            buckets[freq].add(key);
        }

        int resIdx=0;
        for (int i = nums.length;i>=0;i--)
        {
            if (buckets[i]!=null)
                for (int val : buckets[i])
                {
                    res[resIdx++]=val;
                    if (resIdx==k)
                        return res;
                }

        }
return res;
    }
}